### ISSUE TITLE
`"AB" =~ /(A (A|B(*ACCEPT)|C)+ D)(E)/x` doesn't match

### ISSUE THREAD

**Issue by ikegami** [Perl/perl5#19484]
`"AB" =~ /(A (A|B(*ACCEPT)|C)+ D)(E)/x` doesn't match.

---

The following is the example from the documentation for `(*ACCEPT)`:

```
$ perl -e'CORE::say "AB" =~ /(A (A|B(*ACCEPT)|C) D)(E)/x || 0;'
1
```

No surprises there. But if we add a modifier to the alternation...

```
$ perl -e'CORE::say "AB" =~ /(A (A|B(*ACCEPT)|C)+ D)(E)/x || 0;'
0
```

It doesn't even match `ABDE`!

```
$ perl -e'CORE::say "ABDE" =~ /(A (A|B(*ACCEPT)|C)+ D)(E)/x || 0;'
0
```

In both cases, a match isn't even attempted.

```
$ perl -Mre=debug -e'CORE::say "AB" =~ /(A (A|B(*ACCEPT)|C)+ D)(E)/x || 0;'
...
anchored "A" at 0..0 floating "DE" at 1..9223372036854775807 (checking floating) minlen 4
String shorter than min possible regex match (2 < 4)
0
Freeing REx: "(A (A|B(*ACCEPT)|C)+ D)(E)"
```

```
$ perl -Mre=debug -e'CORE::say "ABDE" =~ /(A (A|B(*ACCEPT)|C)+ D)(E)/x || 0;'
...
anchored "A" at 0..0 floating "DE" at 1..9223372036854775807 (checking floating) minlen 4
Matching REx "(A (A|B(*ACCEPT)|C)+ D)(E)" against "ABDE"
Intuit: trying to determine minimum start position...
  doing 'check' fbm scan, [1..3] gave -1
  Did not find floating substr "DE"...
Match rejected by optimizer
0
Freeing REx: "(A (A|B(*ACCEPT)|C)+ D)(E)"
```

---

Perl 5.34.0 built for Ubuntu on WSL using perlbrew.

**Comment by ikegami** [comment]
Source: https://stackoverflow.com/q/71333798/589924

**Comment by demerphq** [comment]
On Thu, 3 Mar 2022 at 09:35, Eric Brine ***@***.***> wrote:

> Source: https://stackoverflow.com/q/71333798/589924
>
>
> Interesting, off the top of my head I would say this is a regression from
when Dave reworked the fixed string logic.

Ill take a look.

Yves

-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

**Comment by demerphq** [comment]
On Thu, 3 Mar 2022 at 11:40, demerphq ***@***.***> wrote:
>
> On Thu, 3 Mar 2022 at 09:35, Eric Brine ***@***.***> wrote:
>>
>> Source: https://stackoverflow.com/q/71333798/589924
>>
>>
> Interesting, off the top of my head I would say this is a regression from when Dave reworked the fixed string logic.

No, I guess this has been broken for a while.

Yves


-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

**Comment by jkeenan** [comment]
> On Thu, 3 Mar 2022 at 11:40, demerphq ***@***.***> wrote: On Thu, 3 Mar 2022 at 09:35, Eric Brine ***@***.***> wrote: > > Source: https://stackoverflow.com/q/71333798/589924 > > Interesting, off the top of my head I would say this is a regression from when Dave reworked the fixed string logic.
> No, I guess this has been broken for a while. Yves

Probably "forever".
```
[~] 519 $ perlbrew use perl-5.34.0
[~] 520 $ perl -e'print "AB" =~ /(A (A|B(*ACCEPT)|C)+ D)(E)/x || 0, "\n";'
0
[~] 521 $ perlbrew use perl-5.10.1
[~] 522 $ perl -e'print "AB" =~ /(A (A|B(*ACCEPT)|C)+ D)(E)/x || 0, "\n";'
0
[~] 523 $ perlbrew use perl-5.8.9
[~] 524 $ perl -e'print "AB" =~ /(A (A|B(*ACCEPT)|C)+ D)(E)/x || 0, "\n";'
Quantifier follows nothing in regex; marked by <-- HERE in m/(A (A|B(* <-- HERE ACCEPT)|C)+ D)(E)/ at -e line 1.
```

**Comment by demerphq** [comment]
On Sun, 13 Mar 2022 at 13:15, James E Keenan ***@***.***>
wrote:

> On Thu, 3 Mar 2022 at 11:40, demerphq *@*.*> wrote: On Thu, 3 Mar 2022 at
> 09:35, Eric Brine @.*> wrote: > > Source:
> https://stackoverflow.com/q/71333798/589924 > > Interesting, off the top
> of my head I would say this is a regression from when Dave reworked the
> fixed string logic.
> No, I guess this has been broken for a while. Yves
>
> Probably "forever".
>
> [~] 519 $ perlbrew use perl-5.34.0
> [~] 520 $ perl -e'print "AB" =~ /(A (A|B(*ACCEPT)|C)+ D)(E)/x || 0, "\n";'
> 0
> [~] 521 $ perlbrew use perl-5.10.1
> [~] 522 $ perl -e'print "AB" =~ /(A (A|B(*ACCEPT)|C)+ D)(E)/x || 0, "\n";'
> 0
> [~] 523 $ perlbrew use perl-5.8.9
> [~] 524 $ perl -e'print "AB" =~ /(A (A|B(*ACCEPT)|C)+ D)(E)/x || 0, "\n";'
> Quantifier follows nothing in regex; marked by <-- HERE in m/(A (A|B(* <-- HERE ACCEPT)|C)+ D)(E)/ at -e line 1.
>
>
Yep. :-)

Yves

-- 
perl -Mre=debug -e "/just|another|perl|hacker/"

**Comment by demerphq** [comment]
HI this should be fixed in 5.36, thanks for the report!
